intel_extension_for_transformers/llm/runtime/graph/models/gptneox/gptneox.cpp

```diff
@@ -135,41 +140,70 @@ static bool gptneox_model_eval_internal(model_context& lctx, const model_token*
-
-        struct ne_tensor* k =
-            ne_view_1d(ctx0, kv_self.k, N * n_embd, (ne_element_size(kv_self.k) * n_embd) * (il * n_ctx + n_past));
-        struct ne_tensor* v =
-            ne_view_2d(ctx0, kv_self.v, N, n_embd, (n_ctx)*ne_element_size(kv_self.v),
-                       (il * n_ctx) * ne_element_size(kv_self.v) * n_embd + n_past * ne_element_size(kv_self.v));
-
-        ne_build_forward_expand(&gf, ne_cpy(ctx0, Kcur, k));
-        ne_build_forward_expand(&gf, ne_cpy(ctx0, Vcur, v));
+        // Vcur = ne_transpose(ctx0, ne_reshape_2d(ctx0, Vcur, n_embd, N * batch_size));
+        std::vector<ne_tensor*> Kcur_bs(batch_size);
+        std::vector<ne_tensor*> Vcur_bs(batch_size);
+        std::vector<ne_tensor*> k_bs(batch_size);
+        std::vector<ne_tensor*> v_bs(batch_size);
+        for (int i = 0; i < batch_size; ++i) {
+          // batch K
+          Kcur_bs[i] = ne_permute(ctx0,
+                                  ne_view_4d(ctx0, Kcur, head_dim, n_head, N, 1, ne_element_size(Kcur) * head_dim,
+                                             ne_element_size(Kcur) * n_embd, ne_element_size(Kcur) * n_embd * N,
+                                             i * ne_element_size(Kcur) * n_embd * N),
+                                  0, 2, 1, 3);
+          k_bs[i] = ne_view_4d(
+              ctx0, kv_self.k, head_dim, N, n_head, 1, ne_element_size(kv_self.k) * head_dim,
+              ne_element_size(kv_self.k) * head_dim * n_ctx, ne_element_size(kv_self.k) * n_embd * n_ctx,
+              ((il * n_ctx) * ne_element_size(kv_self.k) * n_embd * kv_n_ctx_block +
+               i * n_ctx * n_embd * ne_element_size(kv_self.k) + head_dim * n_past * ne_element_size(kv_self.k)));
+
+          // batch V
+          Vcur_bs[i] = ne_permute(ctx0,
+                                  ne_reshape_4d(ctx0,
+                                                ne_view_2d(ctx0, Vcur, n_embd, N, ne_element_size(Vcur) * n_embd,
+                                                           i * ne_element_size(Vcur) * n_embd * N),
+                                                head_dim, n_head, N, 1),
+                                  1, 2, 0, 3);
+          v_bs[i] =
+              ne_view_4d(ctx0, kv_self.v, N, head_dim, n_head, 1, n_ctx * ne_element_size(kv_self.v),
+                         n_ctx * ne_element_size(kv_self.v) * head_dim, n_ctx * ne_element_size(kv_self.v) * n_embd,
+                         ((il * n_ctx) * ne_element_size(kv_self.v) * n_embd * kv_n_ctx_block +
+                          i * n_ctx * n_embd * ne_element_size(kv_self.v) + n_past * ne_element_size(kv_self.v)));
+          ne_build_forward_expand(&gf, ne_cpy(ctx0, Kcur_bs[i], k_bs[i]));
+          ne_build_forward_expand(&gf, ne_cpy(ctx0, Vcur_bs[i], v_bs[i]));
+        }
+
+        // struct ne_tensor* k =
+        //     ne_view_1d(ctx0, kv_self.k, N * n_embd, (ne_element_size(kv_self.k) * n_embd) * (il * n_ctx + n_past));
+        // struct ne_tensor* v =
```

Comment:
remove unuse code

Reply:
fixed in https://github.com/intel/intel-extension-for-transformers/pull/449/commits/540d412fb16d89de60bca98765f4a1e4d231a4e3